File: Majority_Number_III.cpp
class Solution {
public:
    /**
     * @param nums: A list of integers
     * @param k: As described
     * @return: The majority number
     */
    int majorityNumber(vector<int> nums, int k) {
        // write your code here
        //if (occurs / n > 1 / k), then (occusrs - 1)/(n - k) > 1/k
        unordered_map<int, int> hash;
        for (auto& i : nums){
            ++hash[i];
            if (hash.size() == k){
                auto it = hash.begin();
                while (it != hash.end()){
                    if (--it->second == 0){
                        hash.erase(it++);
                    }else{
                        ++it;
                    }
                }
            }
        }
        
        //reset to zero and recount the occurs
        for (auto& it : hash){
            it.second = 0;
        }
        
        for (auto& i : nums){
            auto it = hash.find(i);
            if (it != hash.end()){
                ++it->second;
            }
        }
        //to judge the occurs if > 1/k
        int n = nums.size();
        vector<int> res;
        for (const pair<int, int>& it : hash){
        // for (auto& it : hash){
            if (it.second > static_cast<double>(n) / k){
                res.emplace_back(it.first);
            }
        }
        return res[0];
    }
};

File: Majority Number III.py
class Solution:
    """
    @param nums: A list of integers
    @param k: As described
    @return: The majority number
    """
    def majorityNumber(self, nums, k):
        # write your code here
        
        hashVal = collections.defaultdict(int)
        for n in nums:
            hashVal[n] += 1
            if len(hashVal) == k:
                for key in hashVal.keys():
                    hashVal[key] -= 1
                    if hashVal[key] == 0:
                        hashVal.pop(key)
        
        for key in hashVal.keys():
            hashVal[key] = 0
        for n in nums:
            if n in hashVal:
                hashVal[n] += 1
        
        length = len(nums)
        return [key for key in hashVal.keys() if hashVal[key] > length / k][0]


## ugly
#### using for k in hashV.keys(): and k in arguments will be reseted 
class Solution:
    """
    @param nums: A list of integers
    @param k: As described
    @return: The majority number
    """
    def majorityNumber(self, nums, k):
        # write your code here
        hashV = collections.defaultdict(int)
        for n in nums:
            hashV[n] += 1
            if len(hashV) == k:
                for key in hashV.keys():
                    hashV[key] -= 1
                    if hashV[key] == 0:
                        hashV.pop(key)
                        # break
        # print hashV
        hashV = hashV.fromkeys(hashV.keys(), 0)
        # print hashV
        for n in nums:
            if n in hashV:
                hashV[n] += 1
        # print hashV
        length = len(nums)
        return [key for key in hashV.keys() if hashV[key] * 1.0 / length > 1.0 / k][0]
